Restore service states after install

mattock opened the pull request:
This PR adds NSIS Simple Service Plugin and uses it to

* Get the state of OpenVPNServiceInteractive and OpenVPNService at the beginning of the install
* Restores the states after install (or if installation cannot proceed)
* Sets our sane defaults (enable on boot and start) for OpenVPNServiceInteractive if it was not present (installed) when the install started.

Thus the state of the system remains the same before and after install. This problem was already solved in the OpenVPN Chocolatey package, as discussed in [issue #74](https://github.com/OpenVPN/openvpn-build/issues/74) using Powershell code. It was agreed that implementing this in NSI makes more sense.

The Simple Service Plugin could be used to manage service starting, stopping and configuration also. RIght now we use a mix of "openvpnserv?.exe -install", "net.exe" and "sc.exe" to do service management. I can move all of that to Simple Service Plugin if we want that.

Integration to openvpn.nsi (=this PR) has not yet been tested, but it "should work". Test installers will follow later.

chipitsine commented:
can you please compare sc.exe / net.exe / ?-install /SimplePlugin from the following point of view ... if you run install/uninstall/install, you can see fails because uninstall deletes services in "pending remove" way. I did not notice which of the method leads to that.

we should choose the method which actually removes instead of marking as "pending remove"

mattock commented:
@chipitsine : I'm not sure how to reproduce the PENDING REMOVE issue. Do you get that consistently, or just occasionally?

I fixed a typo (SectionEnd vs. FunctionEnd) and force-pushed. I also switched to using SimpleSC commands for most service control stuff. Based on brief testing things still seem to work as expected. A test installer based on current state of this PR is [here](http://build.openvpn.net/downloads/temp/openvpn-install-2.4.0-I602-pr80v2.exe).

I will now look into replacing the "openvpnserv?.exe -install" calls with SimpleSC.

chipitsine commented:
no, I'm not sure. ok, I'll reproduce by myself later

chipitsine commented:
there's massive eventlog usage, like this

https://github.com/OpenVPN/openvpnserv2/blob/master/Service.cs#L120

I'm afraid we introduced a bug by installing service by sc.exe (and not creating event channel).


what I suggest, is to change logging to standard "Application" event log, as people look at it more often.
or otherwise we need to create event channel during install as well.

P.S. I didn't try to run installer that way (i.e. install without .net 4.0 installed in order to use sc.exe)

mattock commented:
@chipitsine : fortunately sc.exe was only used if .NET 4.0 was not present. You say "massive eventlog usage": do you have sample of an event log with the problem?

Using the Application log makes sense imho, as openvpnserv.exe is already logging in there.

I also wonder how this affects the move to SimpleSC...

mattock commented:
Looking at application event logs I see errors from openvpnserv.exe related to missing registry key. I don't have the full english version of the message, but here's the meat of it:

    Error querying registry value: HKLM\SOFTWARE\OpenVPN\ovpn_admin_group

Example Powershell to get latest full messages:

    Get-EventLog Application -Newest 500|Where-Object { $_.Source -eq "openvpnserv" }|Format-List

This is not related to openvpnserv2, though, but should be fixed regardless.

selvanair commented:
The main difference between using sc and net is that  "net stop", "net start" etc waits for the sate change to complete while sc returns immediately. The SimpleSC plugin will wait (with a timeout) so that way its like using net.exe which is the behaviour we want. 

The issue of "this service has been marked for deletion" (I suppose that's what @chipitsine is referring to?) appear to happen sporadically no matter net is sc is used.  I think its almost always related to something holding up the service from getting fully deleted -- could be task manager, mmc or some such. But not 100% sure.

chipitsine commented:
no, I meant that openvpnserv2 is using different event log, not "Application" by default. it is using it's own log, somewhat like "OpenVpnServ", I do not recall exact name

mattock commented:
More commits added. Test installer v4 [here](http://build.openvpn.net/downloads/temp/openvpn-install-2.4.0-I602-pr80v4.exe).

selvanair commented:
> Looking at application event logs I see errors from openvpnserv.exe related to missing registry key. I don't have the full english version of the message, but here's the meat of it:
> 
> Error querying registry value: HKLM\SOFTWARE\OpenVPN\ovpn_admin_group

I'm aware of this error -- its logged only once at service startup and is caused by the fact that I used existing code in the service to read this registry key. The default behavior of the registry parser in the service is to log an error if a key is not found.

I think the best fix is to add this registry key during installation. That also makes it easier for anyone who wants to customize the group name. 

Apart from that event log entries for iservice appear as from unknown source making it difficult to read the log. Fixing this would be nice.

selvanair commented:
> no, I meant that openvpnserv2 is using different event log, not "Application" by default. it is using it's own log, somewhat like "OpenVpnServ", I do not recall exact name

If that's the case I too think we should change it to log to Application log.

selvanair commented on windows-nsis/openvpn.nsi:
```diff
@@ -244,11 +244,12 @@ Section -pre
 		; Stop & Remove previous OpenVPN service
 		DetailPrint "Removing any previous OpenVPN interactive service..."
-		nsExec::ExecToLog /OEM '"$INSTDIR\bin\openvpnserv.exe" -remove'
+		SimpleSC::RemoveService OpenVPNService
```
Should be "SimpleSC::RemoveService OpenVPNServiceInteractive" isn't it ? 
OpenVPNService has to be removed in the secService if selected.

selvanair commented on windows-nsis/openvpn.nsi:
```diff
@@ -247,7 +247,7 @@ Section -pre
 		; Stop & Remove previous OpenVPN service
 		DetailPrint "Removing any previous OpenVPN interactive and legacy services..."
-		SimpleSC::RemoveService OpenVPNService
+		SimpleSC::RemoveService OpenVPNServiceInteractive
```
Ah, then ignore my previous comment :)

selvanair commented:
One thing seems to be missing: restoring the service starttype. If the user had OpenVPNService set to autostart after upgrade it will revert to manual unless the service starttype is cached and restored. Same dor interactive service  -- if anyone has it set to manual (though they shouldn't)  it will get reverted to auto unless we restore the starttype.

That said, why are we removing the interactive service before upgrade -- just stopping should be enough and then the starttype would be preserved. Same for legacy service. For OpenVPNService we do have to remove it before upgrade as the existing one may be the legacy service left behind from old versions.

chipitsine commented:
I did have a look, actually it is "Application" log. OpenVpnService register new event source

it says "Creating EventLog source OpenVpnService in log Application ..."

we may either move that to "service startup code", i.e. service will perform current installation action (including event source creation) during startup, or we can run something like that during nsis

```
eventcreate /so OpenVpnService /t INFORMATION /id 1 /d OpenVpnService /L Application
```

selvanair commented:
> e may either move that to "service startup code", i.e. service will perform current installation action (including event source creation) during startup, or we can run something like that during nsis
> 
> ```eventcreate /so OpenVpnService /t INFORMATION /id 1 /d OpenVpnService /L Application```

Is this an nsis plugin? If you are referring to eventcreate.exe in windows, that will just write a one-time event to the event log.

mattock commented:
I refactored the service handling code quite a bit in the last commit. A test installer based on latest code is [here](http://build.openvpn.net/downloads/temp/openvpn-install-2.4.0-I602-pr80v5.exe).

The new behavior In a nutshell:
* _OpenVPNServiceLegacy_ and _OpenVPNServiceInteractive_ are no longer uninstalled, only stopped
* _OpenVPNService_ is removed as it might refer to OpenVPNServiceLegacy if upgrading from OpenVPN 2.3.x. 
* State of all previously existing services is always restored
* If a service did not exist before, it gets created with default values, and in the case of _OpenVPNServiceInteractive_, started.
* Service state in this context includes its _status_ (running, stopped) and its _startmode_ (manual, automatic, disabled)

Basically the only ways you should get the new NSI code to change service state is to
* Change a service's startupmode and/or status
* Uninstall OpenVPN (removes all services)
* Install OpenVPN (service no longer exists -> service state reset to default)

I also wrote a [Powershell test script](https://github.com/mattock/openvpn-windows-test/blob/master/Test-Installer.ps1) ([pull request](https://github.com/OpenVPN/openvpn-windows-test/pull/6)) that can be used to ensure that this PR works as intended.

selvanair commented on windows-nsis/openvpn.nsi:
```diff
@@ -344,11 +412,11 @@ Function CoreSetup
-	; install openvpnserv.exe as a services
 	DetailPrint "Installing OpenVPN Interactive Service..."
-	nsExec::ExecToLog /OEM '"$INSTDIR\bin\openvpnserv.exe" -install'
+	SimpleSC::InstallService "OpenVPNServiceInteractive" "OpenVPN Interactive Service" "32" "3" "$INSTDIR\bin\openvpnserv.exe" "tap0901/dhcp" "" ""
```
So we let this error if the service is already present?

mattock replied:
Yes, we can do that. SimpleSC does not produce any output on error, so this is harmless.

selvanair commented on windows-nsis/openvpn.nsi:
```diff
@@ -344,11 +412,11 @@ Function CoreSetup
 
-	Pop $R0 # return value/error/timeout
+	DetailPrint "Installing OpenVPN Legacy Service..."
+	SimpleSC::InstallService "OpenVPNServiceLegacy" "OpenVPN Legacy Service" "32" "3" "$INSTDIR\bin\openvpnserv.exe" "tap0901/dhcp" "" ""
```
Same here..

selvanair commented on windows-nsis/openvpn.nsi:
```diff
@@ -175,6 +175,89 @@ ReserveFile "install-whirl.bmp"
+
+	${IfNot} $service_existed == 0
+		DetailPrint "Setting starttype of OpenVPN Service to manual"
+		SimpleSC::SetServiceStartType "OpenVPNService" 3
```
This too looks redundant as its set during install

selvanair commented on windows-nsis/openvpn.nsi:
```diff
@@ -175,6 +175,89 @@ ReserveFile "install-whirl.bmp"
+
+	${IfNot} $legacy_service_existed == 0
+		DetailPrint "Setting starttype of OpenVPN Legacy Service to manual"
+		SimpleSC::SetServiceStartType "OpenVPNServiceLegacy" 3
```
Same here: if not existed, starttype is already set during install. But reinstall to new location requires a binary_path reset.

selvanair commented on windows-nsis/openvpn.nsi:
```diff
@@ -175,6 +175,89 @@ ReserveFile "install-whirl.bmp"
+	${IfNot} $iservice_existed == 0
+		DetailPrint "Enabling and starting OpenVPN Interactive service" 
+		SimpleSC::SetServiceStartType "OpenVPNServiceInteractive" 2
+		SimpleSC::StartService "OpenVPNServiceInteractive" "" 5
```
The service StartType in this case is already set by the call to install, so is not needed ?

If the user is installing to a new INSTDIR without uninstalling the previous one we also have to reset the service executable path. So need an ${Else}  SimpleSC::SetServiceBinaryPath .... ?

mattock commented:
I'll look into the redundant StartType and binary path issues early next week.

mattock commented:
I refactored the code quite heavily and squashed all the commits together. I used the Powershell script abaove to ensure that the logic works as expected. I believe all of Selvanair's concerns have been addressed now.

I also found and fixed a bug that was masked before. The problem was that _running_ services would not get removed in the Uninstall section by SimpleSC::RemoveService. These services would then be left behind to cause odd side-effects during subsequent installs.

A test installer is available [here](https://build.openvpn.net/downloads/snapshots/openvpn-install-2.4.0-I602-pr80v18.exe). Let me know if further changes are required.

selvanair commented:
Looks good and works as promised. One question:
Any reason why  interactive service gets only 5 sec timeout for start/stop while the other two are given 10 secs? As all three are stopped together, cumulatively it seems to get ample time to stop, so just being curious..

ACK.

mattock commented:
@selvanair : I assumed that OpenVPNService and OpenVPNServiceLegacy need more time as they have to shut down all the OpenVPN instances they're managing. That said, I don't know if all the OpenVPN instances have to actually be killed before the service is marked as "stopped". I'll increase the start/stop timeout for OpenVPNServiceInteractive to 10 seconds, force-push and then merge this.

chipitsine commented:
how those timeouts are implemented in SimpleSC ?
does it wait for 10 sec and checks for service state ? or does it check every second ?

mattock commented:
Looks like SimpleSC polls the service status every 250 milliseconds. These polls continue until the desired status is reached. It exists with ERROR_SERVICE_REQUEST_TIMEOUT if timeout is reached before this happens.

chipitsine commented:
wow

mattock commented:
Right now the exit code of SimpleSC commands are not displayed in the install log, which could be problematic when diagnosing problems. I'll see if this could be fixed without adding tons of new code.

mattock commented:
I increased OpenVPNServiceInteractive start/stop timeout to 10 seconds.

Printing exit codes of the SimpleSC commands would increase patch size considerably. I'll merge this as-is, and we can add debugging to places where people encounter issues (if any).

wget commented:
Sometimes, I wonder if that wasn't easier to leverage msi Windows Installer files. Have just read tutorials, this isn't too complicated to code.

mattock commented:
@wget : I don't particularly like NSIS as it feels like a big mess. That said, it works very reliably, and more importantly has never prevented us from doing what we need. Most of what our installer does is basic stuff (e.g. files and registry keys) , but based on a quick look there are some exceptions:

* Offers to install .NET 4.5 framework if it is missing (unless install is silent)
* Caches/restores service states
* Refuses to start the install if (manually launched) openvpn.exe instances are running
* Runs the tap-windows6 installer from within the openvpn installer with correct options
* (Injects the publisher certificate to the certificate store)

These boil down to running some commands, processing command output and acting accordingly if necessary. Any ideas how such things would be handled using MSI?

chipitsine commented:
I beleive such things could be handled by pull requests :)
and, it looks that there's new developer in openvpn area!

wget commented:
@mattock  +1 NSIS for the mess. Btw, msi offers the advantage to be deployed using the Enterprise Windows Update and using GPO. This is why is should be privileged on Windows environments.

Without being 100% certain, I think msi could meet the aforementioned requirements.

I'm gonna look at this as soon I catched up with the other chocolatey packages I manage (the purpose is to merge them in [coreteampackages](https://github.com/wget/chocolatey-coreteampackages/)) and have finished the code required to have a snapshot of the OpenVPN Chocolatey package (cf. [issue#1](https://github.com/wget/chocolatey-package-openvpn/issues/2)). I begin to become fluent with AppVeyor, but this take time as my free time is limited :)
